engine/src/flutter/tools/pkg/engine_build_configs/bin/check.dart

```diff
@@ -39,36 +94,112 @@ void main(List<String> args) {
   }
-  if (loader.errors.isNotEmpty) {
-    loader.errors.forEach(io.stderr.writeln);
-    io.exitCode = 1;
+  indentedPrint(loader.errors);
+
+  // Find and parse the .ci.yaml configuration (for the engine).
+  final CiConfig? ciConfig;
+  {
+    final String ciYamlPath = p.join(engine.flutterDir.path, '.ci.yaml');
+    final String realCiYaml = io.File(ciYamlPath).readAsStringSync();
+    final y.YamlNode yamlNode = y.loadYamlNode(realCiYaml, sourceUrl: Uri.file(ciYamlPath));
+    final loadedConfig = CiConfig.fromYaml(yamlNode);
+
+    statusPrint('.ci.yaml at ${p.relative(ciYamlPath)} is valid', success: loadedConfig.valid);
+    if (!loadedConfig.valid) {
+      indentedPrint([loadedConfig.error!]);
+      ciConfig = null;
+    } else {
+      ciConfig = loadedConfig;
+    }
   }
 
   // Check the parsed build configs for validity.
   final List<String> invalidErrors = checkForInvalidConfigs(configs);
-  if (invalidErrors.isNotEmpty) {
-    invalidErrors.forEach(io.stderr.writeln);
-    io.exitCode = 1;
-  }
+  statusPrint('All configuration files are valid', success: invalidErrors.isEmpty);
+  indentedPrint(invalidErrors);
 
   // We require all builds within a builder config to be uniquely named.
   final List<String> duplicateErrors = checkForDuplicateConfigs(configs);
-  if (duplicateErrors.isNotEmpty) {
-    duplicateErrors.forEach(io.stderr.writeln);
-    io.exitCode = 1;
-  }
+  statusPrint('All builds within a builder are uniquely named', success: duplicateErrors.isEmpty);
+  indentedPrint(duplicateErrors);
 
   // We require all builds to be named in a way that is understood by et.
   final List<String> buildNameErrors = checkForInvalidBuildNames(configs);
-  if (buildNameErrors.isNotEmpty) {
-    buildNameErrors.forEach(io.stderr.writeln);
-    io.exitCode = 1;
+  statusPrint('All build names must have a conforming prefix', success: buildNameErrors.isEmpty);
+  indentedPrint(buildNameErrors);
+
+  // If we have a successfully parsed .ci.yaml, perform additional checks.
+  if (ciConfig == null) {
+    return;
+  }
+
+  // We require that targets that have `properties: release_build: "true"`:
+  // (1) Each sub-build produces artifacts (`archives: [...]`)
+  // (2) Each sub-build does not have `tests: [ ... ]`
+  final buildConventionErrors = <String>[];
+  for (final MapEntry(key: _, value: target) in ciConfig.ciTargets.entries) {
+    final config = loader.configs[target.properties.configName];
+    if (target.properties.configName == null) {
+      // * builder_cache targets do not have configuration files.
+      debugPrint('  Skipping ${target.name}: No configuration file found');
+      continue;
+    }
+
+    // This would fail above during the general loading.
+    if (config == null) {
+      throw StateError('Unreachable');
+    }
+
+    final configConventionErrors = <String>[];
+    if (target.properties.isReleaseBuilder) {
+      // If there is a global generators step, assume it is doing something unscrupulous.
```

Comment:
maybe instead: assume artifacts are uploaded from the generators

Reply:
Done!